components/server/src/authorization/spicedb-authorizer.ts

```diff
@@ -196,3 +228,111 @@ export class SpiceDBAuthorizerImpl implements SpiceDBAuthorizer {
+ */
+export class RequestLocalZedTokenCache implements ZedTokenCache {
+    constructor() {}
+
+    async get(objectRef: v1.ObjectReference | undefined): Promise<string | undefined> {
+        return getContext().zedToken?.token;
+    }
+
+    async set(...kvs: ZedTokenCacheKV[]) {
+        function clearZedTokenOnContext() {
+            getContext().zedToken = undefined;
+        }
+
+        const mustClearCache = kvs.some(([k, v]) => !!k && !v); // did we write a relationship without getting a writtenAt token?
+        if (mustClearCache) {
+            clearZedTokenOnContext();
+            return;
+        }
+
+        try {
+            const allTokens = [
+                ...kvs.map(([_, v]) => (!!v ? StoredZedToken.fromToken(v) : undefined)),
+                getContext().zedToken,
+            ].filter((v) => !!v) as StoredZedToken[];
+            const freshest = this.freshest(...allTokens);
+            if (freshest) {
+                getContext().zedToken = freshest;
+            }
+        } catch (err) {
+            log.warn("[spicedb] Failed to set ZedToken on context", err);
+            clearZedTokenOnContext();
+        }
+    }
+
+    async consistency(resourceRef: v1.ObjectReference | undefined): Promise<v1.Consistency> {
+        function fullyConsistent() {
+            return v1.Consistency.create({
+                requirement: {
+                    oneofKind: "fullyConsistent",
+                    fullyConsistent: true,
+                },
+            });
+        }
+
+        const zedToken = await this.get(resourceRef);
+        if (!zedToken) {
+            return fullyConsistent();
+        }
+        return v1.Consistency.create({
+            requirement: {
+                oneofKind: "atLeastAsFresh",
+                atLeastAsFresh: v1.ZedToken.create({
+                    token: zedToken,
+                }),
+            },
+        });
+    }
+
+    protected freshest(...zedTokens: StoredZedToken[]): StoredZedToken | undefined {
+        return zedTokens.reduce<StoredZedToken | undefined>((prev, curr) => {
+            if (!prev || prev.timestamp < curr.timestamp) {
+                return curr;
+            }
+            return prev;
+        }, undefined);
+    }
+}
+
+export interface StoredZedToken {
+    token: string;
+    timestamp: number;
+}
+namespace StoredZedToken {
+    export function create(token: string, timestamp: number): StoredZedToken {
+        return { token, timestamp };
+    }
+
+    export function fromToken(token: string): StoredZedToken | undefined {
+        // following https://github.com/authzed/spicedb/blob/786555c24af98abfe3f832c94dbae5ca518dcf50/pkg/zedtoken/zedtoken.go#L64-L100
+        const decodedBytes = base64decode(token);
+        const decodedToken = DecodedZedToken.decode(Buffer.from(decodedBytes, "utf8")).v1;
+        if (!decodedToken) {
```

Comment:
is that not an error case?

Reply:
Yes, this is a technical error. But it would not break in a user-facing way - except we got a bunch of these, which we need to detect via metrics+alerts.

I think we need to decide (and document) how we see log levels being used, because I heart people (over time) argue for very different perspectives. :+1:

Reply:
Probably similar case to the above. I think we should at least log an error